### PR TITLE
Feat/in memory database

### LINKED ISSUE #26: Implement Storage for Statistics Data with In-Memory Object and Periodic Serialization
# Top priority
# Task: Develop Storage Solution for Statistics Data  

## **Objective**  
The goal of this task is to develop a storage solution for statistics data that:  
1. Maintains an **in-memory object** for real-time updates and retrieval.  
2. Periodically serializes the in-memory data into a **JSON file** for persistence.  
3. Provides an **interface** to use this storage.  

---

#### **1.1 JsonSerializable Interface**  
Create a `JsonSerializable` interface that provides:  
- A function for JSON serialization of a class.  
- A function that returns the name of the class in JSON format.  

#### **1.2 In-Memory Storage**  
- Create a **service class** to hold the in-memory object.  
- Provide methods to:  
  - Add, update, retrieve, and delete data in the in-memory object.  
- Ensure methods work for classes implementing the `JsonSerializable` interface.  

#### **1.3 Periodic Serialization/Deserialization**  
- Use the `@Scheduled` annotation to schedule periodic tasks that serialize the in-memory object into a JSON file (e.g., every 10 minutes).  
- Save the JSON file to a configurable location (e.g., `/data/statistics.json`) defined by a parameter in the properties file.  
- The JSON object should consist of:  
  - The name of the class.  
  - The serialized content of the class.  
- Implement deserialization to load data from the JSON file back into the in-memory object at application startup.

---

#### **Sample Class**  
- Create a sample class that implements the `JsonSerializable` interface.  

#### **Unit Tests**  
- Test the **addition**, **update**, and **retrieval** of data in the in-memory object.  
- Verify the correct serialization of the in-memory object to a JSON file.  

#### **Edge Cases**  
- Test behavior when:  
  - The JSON object is missing.  
  - The JSON object is malformed or contains incomplete data.  

---

### **3. Documentation**  
Provide developer documentation that explains:  
- How to implement the `JsonSerializable` interface.  
- The interface for using this storage.  

---

### **1. Functional Implementation**  
- An interface is available with two methods:  
  - A function for serializing a class instance to JSON.  
  - A function for retrieving the name of the class in JSON format.  

- A **thread-safe service class** is implemented for managing in-memory objects of classes implementing `JsonSerializable`.  
- The service provides methods to **add**, **update**, **retrieve**, and **delete** data for any `JsonSerializable` object.  
- The in-memory storage is flexible and can handle various types of data that implement the `JsonSerializable` interface.  

---

### **2. Periodic Serialization**  
- Data in the in-memory storage is serialized into a JSON file every 10 minutes.  
- The JSON object includes:  
  - The name of the class.  
  - The serialized content.  
- The file is saved to a configurable location specified in the application's properties file (e.g., `/data/statistics.json`).  

---

### **3. Robust Testing**  
- Unit tests ensure the correct **addition**, **update**, **retrieval**, and **deletion** of data in the in-memory storage.  
- Tests verify accurate serialization of the in-memory object into the expected JSON format.  
- Tests validate the system's behavior for edge cases, such as:  
  - The JSON file is missing.  
  - The JSON file is malformed or contains incomplete data.  
- The system gracefully handles errors and provides appropriate error messages or fallback mechanisms.  

---

### **4. Clear Documentation**  
Developer documentation is provided and includes:  
- A detailed explanation of the `JsonSerializable` interface, including its purpose and methods.  
- Instructions for using the in-memory storage service, with examples for:  
  - Adding, updating, retrieving, and deleting data.  
- Guidelines for configuring the JSON file's save location and scheduling parameters.  
- Steps for integrating new classes with the storage by implementing the `JsonSerializable` interface.

### PR DESCRIPTION
**Pull request by dominikkorwek**
I will add documentation after fixes

### CODE REVIEW

File: src/main/java/pl/wrapper/parking/facade/SerializerService.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package pl.wrapper.parking.facade;
+
+import pl.wrapper.parking.infrastructure.error.Result;
+
+import java.io.Serializable;
+import java.util.List;
+
+public interface SerializerService<T extends Serializable> {
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/facade/SerializerService.java]: I'm not exactly sure what this is and whether we need it

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepository.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package pl.wrapper.parking.infrastructure.inMemory;
+
+import java.io.Serializable;
+
+public interface InMemoryRepository<K extends Serializable, V extends Serializable> {
+
+    void add(K key, V value);
+    V get(K key);
+    //add a method to fetch all k,v pairs from dataMap
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepository.java]: add a method to fetch all k,v pairs from dataMap

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.springframework.core.serializer.support.SerializationFailedException;
+import org.springframework.scheduling.annotation.Scheduled;
+
+import java.io.*;
+import java.util.Map;
+
+public abstract class InMemoryRepositoryImpl<K extends Serializable, V extends Serializable> implements InMemoryRepository<K, V> {
+    protected final transient File file;
+    protected Map<K, V> dataMap;
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java:16]: let's change List to map. Using indexes is a bad idea - especially since they rely on the order of adding the data, which can change

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    protected V defaultValue;
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java:17]: Let's add a default value

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    @PreDestroy
+    private void selfSerialize() {
+        File parent = file.getParentFile();
+        if (parent != null && !parent.exists())
+            if (!parent.mkdirs())
+                throw new SerializationFailedException("Failed to create directory for path: " + file.getAbsolutePath());
+        try {
+            FileOutputStream fileOut = new FileOutputStream(file);
+            ObjectOutputStream out = new ObjectOutputStream(fileOut);
+            out.writeObject(this);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java]: do not serialize whole, let's serialize only the dataMap. Should be far easier that way

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            //do not serialize whole, let's serialize only the dataMap. Should be far easier that way
+        } catch (IOException e) {
+            throw new SerializationFailedException(createExceptionForIOE("Serialization", e));
+        }
+        //Add handling for these exceptions in exceptionHandler
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java]: Add handling for these exceptions in exceptionHandler

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    protected  <E extends Exception> String createExceptionForIOE(String methodType, E e) {
+        //make it static
+        return methodType + " failed for: " + this.getClass().getSimpleName() + ". Message: " + e.getMessage();
+    }
+
+    @Override
+    public void add(K key, V value) {
+        //add here
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java]: add implementation for this

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Scheduled(fixedRateString = "#{60 * 100 * ${serialization.timeStamp.inMinutes}}")
+    private void periodicSerialize() {
+        selfSerialize();
+    }
+
+    protected  <E extends Exception> String createExceptionForIOE(String methodType, E e) {
+        //make it static
+        return methodType + " failed for: " + this.getClass().getSimpleName() + ". Message: " + e.getMessage();
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java]: Let's make this static. Also we don't need to make it a template method

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.core.serializer.support.SerializationFailedException;
+import org.springframework.stereotype.Component;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.util.HashMap;
+
+@Component
+public class ParkingDataRepository extends InMemoryRepositoryImpl<Integer, SomeObject> { //K and V types can/should be changed, but leave that for those who have the other two tasks
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java]: create another class like this to store ParkingEndpoint data. You only need to implement the constructor and init method

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    public ParkingDataRepository(@Value("${serialization.location.parkingData}") String saveToLocationPath) {
+        super(saveToLocationPath);
+        this.defaultValue = null; //Add default value here (empty object probably), but leave that for those who have the other two tasks
+    }
+
+    @Override
+    protected void init() {
+        if (!file.exists()) {
+            this.dataMap = new HashMap<>(); //put here whatever map type you want
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java]: Leave choosing the K/V, default value impl and dataMap type  for those who the other two tasks. Leave a helpful comment or two for them so they know what should be changed

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            return;
+        }
+
+        try (FileInputStream fileOut = new FileInputStream(file);
+             ObjectInputStream in = new ObjectInputStream(fileOut)) {
+
+//            InMemoryRepoBasic<V> deserializedObject = (InMemoryRepoBasic<V>) in.readObject();
+//            this.data = deserializedObject.data;
+            //change it (we will serialize and read only the dataMap, not the whole object)
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java]: Since we are serializing only the dataMap, this code should be updated

File: src/main/resources/application.properties [hunk 12]
```diff
@@ -5,5 +5,12 @@ server.port=8080
 
 maps.api.url=https://nominatim.openstreetmap.org
 
+serialization.timeStamp.inMinutes=10
+
+serialization.location=data/statistics
+serialization.location.basic =${serialization.location}/basic
```
Comment (Leadman5555) [inline src/main/resources/application.properties]: add property for parkingData and parkingEnpoints, remove the basic one, we don't need it

File: src/test/java/pl/wrapper/parking/InMemory/InMemoryObjectTest.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
+//package pl.wrapper.parking.InMemory;
+//
+//import java.io.Serializable;
+//
+//@SuppressWarnings("unused")
+//public class InMemoryObjectTest implements Serializable {
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/InMemory/InMemoryObjectTest.java]: Why is a test thing, whatever this is, in a non-test package?

File: src/test/java/pl/wrapper/parking/InMemory/InMemoryServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+//
+//import org.springframework.stereotype.Service;
+//import pl.wrapper.parking.facade.SerializerService;
+//import pl.wrapper.parking.infrastructure.error.Result;
+//import pl.wrapper.parking.infrastructure.inMemory.InMemoryRepo;
+//
+//import java.util.List;
+//
+//@Service
+//public record InMemoryServiceTest(InMemoryRepo<InMemoryObjectTest> repo) implements SerializerService<InMemoryObjectTest> {
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/InMemory/InMemoryServiceTest.java]: Again, why is a test thing in a non-test package?

File: src/test/java/pl/wrapper/parking/InMemory/InMemoryServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
+//package pl.wrapper.parking.InMemory;
+//
+//import org.springframework.stereotype.Service;
+//import pl.wrapper.parking.facade.SerializerService;
+//import pl.wrapper.parking.infrastructure.error.Result;
+//import pl.wrapper.parking.infrastructure.inMemory.InMemoryRepo;
+//
+//import java.util.List;
+//
+//@Service
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/InMemory/InMemoryServiceTest.java]: A Test service bean is a very bad idea. Try to think of a different way do this

File: src/test/java/pl/wrapper/parking/InMemory/InMemoryTest.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+//import lombok.extern.slf4j.Slf4j;
+//import org.junit.jupiter.api.Test;
+//import org.springframework.beans.factory.annotation.Autowired;
+//import org.springframework.boot.test.context.SpringBootTest;
+//
+//
+//import static org.junit.jupiter.api.Assertions.*;
+//
+//@SpringBootTest
+//@Slf4j
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/InMemory/InMemoryTest.java]: Adding logger to tests is also a bad idea, let's remove it

File: src/test/java/pl/wrapper/parking/InMemory/InMemoryTest.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+//
+//import lombok.extern.slf4j.Slf4j;
+//import org.junit.jupiter.api.Test;
+//import org.springframework.beans.factory.annotation.Autowired;
+//import org.springframework.boot.test.context.SpringBootTest;
+//
+//
+//import static org.junit.jupiter.api.Assertions.*;
+//
+//@SpringBootTest
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/InMemory/InMemoryTest.java]: Double check if we really need the full spring context here. If we don't, change this annotation.

File: src/test/java/pl/wrapper/parking/infrastructure/error/ResultTest.java [hunk 18]
```diff
@@ -1,58 +1,58 @@
...
+//import org.springframework.test.web.servlet.MockMvc;
+//import org.springframework.test.web.servlet.MvcResult;
+//import pl.wrapper.parking.facade.domain.DummyController;
+//import pl.wrapper.parking.facade.domain.DummyService;
+//import pl.wrapper.parking.pwrResponseHandler.PwrApiServerCaller;
+//import pl.wrapper.parking.pwrResponseHandler.domain.PwrApiCaller;
+//import pl.wrapper.parking.pwrResponseHandler.domain.PwrApiServerCallerImpl;
+//import reactor.core.publisher.Mono;
+//
+//@WebMvcTest(DummyController.class)
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/infrastructure/error/ResultTest.java]: Refactor the result test. We no longer need the dummyController and Service. Use the already existing errors

File: src/test/java/pl/wrapper/parking/InMemory/InMemoryTest.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+//@Slf4j
+//public class InMemoryTest {
+//
+//    @Autowired
+//    private InMemoryServiceTest serviceTest;
+//
+//
+//    @Test
+//    public void shouldReturnData() {
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/InMemory/InMemoryTest.java]: No unit test should require you to manually check outside files. Let's change this implementation to avoid doing that

Comment (Leadman5555) [reply src/test/java/pl/wrapper/parking/InMemory/InMemoryTest.java]: Also, we should add more tests here. Especially the serialization and deserialization cycle.

**Review by Leadman5555 — CHANGES_REQUESTED** [review]
Some changes needed. Use the updated code version after my last commit

File: src/test/java/pl/wrapper/parking/infrastructure/inMemory/ParkingEndpointRepositoryTest.java [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+        int first = inMemoryRepository.fetch().stream().findFirst().orElseThrow();
+
+        assertEquals(first, id);
+
+        assertEquals(inMemoryRepository.get(first), object);
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void shouldSerializationCircleRunCorrectly() throws NoSuchMethodException, InvocationTargetException, IllegalAccessException, NoSuchFieldException {
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/infrastructure/inMemory/ParkingEndpointRepositoryTest.java]: Change this method name, "CircleRun" sounds... weird.

File: src/test/java/pl/wrapper/parking/infrastructure/inMemory/ParkingEndpointRepositoryTest.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
+package pl.wrapper.parking.infrastructure.inMemory;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+
+import java.io.File;
+import java.lang.reflect.Field;
```
Comment (Leadman5555) [inline src/test/java/pl/wrapper/parking/infrastructure/inMemory/ParkingEndpointRepositoryTest.java]: Do not use Reflect. It's never a good idea and it disrupts the test flow. Why are we even using it, by the way? There is no need to.

File: src/main/resources/application.properties [hunk 22]
```diff
@@ -5,5 +5,11 @@ server.port=8080
 
 maps.api.url=https://nominatim.openstreetmap.org
 
+serialization.timeStamp.inMinutes=10
+
```
Comment (Leadman5555) [inline src/main/resources/application.properties:9]: Add application properties file to tests, copy those properties there.

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingEndpointRepository.java [hunk 23]
```diff
@@ -0,0 +1,15 @@
+package pl.wrapper.parking.infrastructure.inMemory;
+
+import org.springframework.beans.factory.annotation.Value;
+
+import java.util.HashMap;
+
+public class ParkingEndpointRepository extends InMemoryRepositoryImpl<Integer, DummyObject> {  // <Key, value>
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingEndpointRepository.java]: Add @Component annotation here

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+            ObjectOutputStream out = new ObjectOutputStream(fileOut);
+
+            out.writeObject(dataMap);
+        } catch (IOException e) {
+            throw new SerializationFailedException(createExceptionForIOE("Serialization", e));
+        }
+    }
+
+    @Scheduled(fixedRateString = "#{60 * 100 * ${serialization.timeStamp.inMinutes}}")
+    private void periodicSerialize() {
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java]: Make this method protected

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+        File parent = file.getParentFile();
+        if (parent != null && !parent.exists())
+            if (!parent.mkdirs())
+                throw new SerializationFailedException(
+                        "Failed to create directory for path: " + file.getAbsolutePath());
+
+        try (FileOutputStream fileOut = new FileOutputStream(file)) {
+            ObjectOutputStream out = new ObjectOutputStream(fileOut);
+
+            out.writeObject(dataMap);
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java:71]: Close the stream here too

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public V get(K key) {
+        return dataMap.getOrDefault(key, defaultValue);
+    }
+
+    @PostConstruct
+    @SuppressWarnings("unchecked")
+    protected void init() {
+        try (FileInputStream fileOut = new FileInputStream(file);
+                ObjectInputStream in = new ObjectInputStream(fileOut)) {
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java:52]: Let's close the stream, it's good practice to, even if not necessary

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+    public InMemoryRepositoryImpl(String filePath, Map<K, V> map, V defaultValue) {
+        this.file = new File(filePath);
+        this.defaultValue = defaultValue;
+
+        this.dataMap = map;
+    }
+
+    @Override
+    public void add(K key, V value) {
+        if (key == null || value == null)
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepositoryImpl.java]: We don't really need it. I mean, we can trust people using this method not to do that, right...?

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepository.java [hunk 28]
```diff
@@ -0,0 +1,10 @@
+package pl.wrapper.parking.infrastructure.inMemory;
+
+import java.io.Serializable;
+import java.util.Set;
+
+public interface InMemoryRepository<K extends Serializable, V extends Serializable> {
+    void add(K key, V value);
+    V get(K key);
+    Set<K> fetch();
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/InMemoryRepository.java]: 1. Change method name to fetchAllKeys().
2. Add a method fetchAllEntries that returns a a Set of Entries.

File: src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java [hunk 29]
```diff
@@ -0,0 +1,16 @@
+package pl.wrapper.parking.infrastructure.inMemory;
+
+import java.util.HashMap;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ParkingDataRepository extends InMemoryRepositoryImpl<Integer, DummyObject> { // <Key, value>
```
Comment (Leadman5555) [inline src/main/java/pl/wrapper/parking/infrastructure/inMemory/ParkingDataRepository.java:8]: It's good practice to name custom components, add name like "ParkingDataRepository"

**Review by Leadman5555 — CHANGES_REQUESTED** [review]
Before sending a pull request, make sure the application can boot up. Currently, it can't - exception when trying to read from file: 
"Caused by: org.springframework.core.serializer.support.SerializationFailedException: Deserialization failed for: InMemoryRepositoryImpl. Message: data\statistics\data (The system cannot find the path specified)".
Fix it before addressing other requested changes.

**Review by Leadman5555 — APPROVED** [review]
